src/client/datascience/errors/errorHandler.ts

```diff
@@ -53,10 +63,20 @@ import { JupyterInterpreterService } from '../jupyter/interpreter/jupyterInterpr
 import { ProductNames } from '../../common/installer/productNames';
 import { EnvironmentType } from '../../pythonEnvironments/info';
 import { JupyterInterpreterDependencyResponse } from '../jupyter/interpreter/jupyterInterpreterDependencyService';
-import { translateProductToModule } from '../../common/installer/productInstaller';
+import {
+    clearInstalledIntoInterpreterMemento,
+    translateProductToModule
+} from '../../common/installer/productInstaller';
+import { JupyterInvalidKernelError } from './jupyterInvalidKernelError';
+import { selectKernel } from '../jupyter/kernels/kernelSelector';
 
+function getFirstCell(cells?: NotebookCell[]) {
+    return cells?.length ? cells[0] : undefined;
+}
 @injectable()
 export class DataScienceErrorHandler implements IDataScienceErrorHandler {
+    private readonly _onShouldRunCells = new EventEmitter<NotebookCell[]>();
+    public readonly onShouldRunCells = this._onShouldRunCells.event;
```

Comment:
Error handler handles the errors now and determines the actions to be taken.
E.g. if ipykernel is not installed, error handler will check if user needs to select a new kernel and once done will then notify other parts of extension whether cells need to be re-run.

Reply:
I did this another way. Make the errorHandler return a specific result that says a switch happened. I'll see what effect this has on my change.